linkerd/meshtls/boring/src/client.rs

```diff
@@ -0,0 +1,163 @@
+use boring::ssl;
+use linkerd_error::Result;
+use linkerd_identity::Name;
+use linkerd_io as io;
+use linkerd_stack::{NewService, Service};
+use linkerd_tls::{
+    client::AlpnProtocols, ClientTls, HasNegotiatedProtocol, NegotiatedProtocolRef, ServerId,
+};
+use std::{future::Future, pin::Pin};
+use tokio::sync::watch;
+
+#[derive(Clone)]
+pub struct NewClient(watch::Receiver<ssl::SslConnector>);
+
+#[derive(Clone)]
+pub struct Connect {
+    server_id: Name,
+    connector: ssl::SslConnector,
+}
+
+pub type ConnectFuture<I> = Pin<Box<dyn Future<Output = io::Result<ClientIo<I>>> + Send>>;
+
+#[derive(Debug)]
+pub struct ClientIo<I>(tokio_boring::SslStream<I>);
+
+// === impl NewClient ===
+
+impl NewClient {
+    pub(crate) fn new(rx: watch::Receiver<ssl::SslConnector>) -> Self {
+        Self(rx)
+    }
+}
+
+impl NewService<ClientTls> for NewClient {
+    type Service = Connect;
+
+    fn new_service(&self, target: ClientTls) -> Self::Service {
+        Connect::new(target, (*self.0.borrow()).clone())
+    }
+}
+
+impl std::fmt::Debug for NewClient {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("NewClient").finish()
+    }
+}
+
+// === impl Connect ===
+
+impl Connect {
+    pub(crate) fn new(client_tls: ClientTls, connector: ssl::SslConnector) -> Self {
+        let ServerId(server_id) = client_tls.server_id;
+
+        if let Some(AlpnProtocols(protocols)) = client_tls.alpn {
+            if !protocols.is_empty() {
+                todo!("support ALPN")
+            }
+        }
+
+        Self {
+            server_id,
+            connector,
+        }
+    }
+}
+
+impl<I> Service<I> for Connect
+where
+    I: io::AsyncRead + io::AsyncWrite + Send + Sync + Unpin + 'static,
+{
+    type Response = ClientIo<I>;
+    type Error = io::Error;
+    type Future = ConnectFuture<I>;
+
+    fn poll_ready(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        std::task::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        let conn = self.connector.clone();
+        let id = self.server_id.clone();
+        Box::pin(async move {
+            let config = conn
+                .configure()
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+            let io = tokio_boring::connect(config, id.as_str(), io)
+                .await
+                .map_err(|e| match e.as_io_error() {
+                    Some(ioe) => io::Error::new(ioe.kind(), ioe.to_string()),
```

Comment:
wait...what's going on here? presumably, `as_io_error()` returns an `Option<io::Error>`; why do we create a new `io::Error` rather than just returning that error?

Reply:
...oh, I see, `as_io_error()` returns a reference, and there's no way to get the inner `io::Error` out of the `HandshakeError`. Gross. That could be worth an upstream PR?

Reply:
Yeah this is definitely not ideal :(
Most definitely would warrant and upstream PR with some changes to the error handling for handshake errors.

Reply:
i went ahead and opened https://github.com/cloudflare/boring/pull/57 upstream, which should hopefully help with this!

Reply:
I think stringifying the underlying error here is fine for now. I looked at submitting a PR and it's a little more complicated than I expected. If someone else wants to submit an upstream change, we'll use it; but I'm not sure it's worth the effort.